### PR TITLE
mgr/cephadm: asyncio based universal timeout for ssh/cephadm commands

### PR DESCRIPTION
**Pull request by adk3798**
Since we already have make use of asyncio for our ssh commands,
we can use asyncio's timeout on waiting for concurrent futures to complete
as a way to have universal timeouts on our cephadm commands.
This change also creates a contextmanager that will catch any asyncio.TimeoutError.
Using the contextmanager along with calls to the wait_async function
will catch any timeout exception raised and convert it into an appropriate
OrchetratorError including information about what and where for the timeout
if it was provided (host where run, what command). This allows us to guarantee a
background ssh command eventually returns and inform users of any
timeouts by raising a health warning or logging the error instead
of sitting idle indefinitely

Fixes: https://tracker.ceph.com/issues/54024

Signed-off-by: Adam King <adking@redhat.com>

---

Still need to add a brief docs section for the new config option to control the timeout

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/tests/test_cephadm.py [hunk 1]
```diff
@@ -2261,3 +2262,33 @@ def test_inventory_known_hostnames(self, cephadm_module):
...
+        async def _timeout():
+            raise asyncio.TimeoutError
+
+        with pytest.raises(OrchestratorError, match=r'Command timed out \(default 900 second timeout\)'):
+            with cephadm_module.async_timeout_handler():
+                cephadm_module.wait_async(_timeout())
+
+        with pytest.raises(OrchestratorError, match=r'Command timed out on host hostA \(default 900 second timeout\)'):
+            with cephadm_module.async_timeout_handler('hostA'):
+                cephadm_module.wait_async(_timeout())
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/tests/test_cephadm.py:2278]: Just to clarify: are we really testing something here rather than passing the parameters?

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-04-04_02:06:27-orch:cephadm-wip-adk-testing-2023-04-03-1541-distro-default-smithi/

Reruns of failed/dead jobs:  https://pulpito.ceph.com/adking-2023-04-04_05:19:12-orch:cephadm-wip-adk-testing-2023-04-03-1541-distro-default-smithi/

After reruns, 6 failures. All 6 failures were on tests being added/adjusted by PRs in the run (2 keepalive-only nfs tests, 2 mon set crush location tests, 2 staggered upgrade tests) and those failures don't seem like they should block merging for any PRs in the run outside of those adjusting the tests.

File: src/pybind/mgr/cephadm/tests/test_cephadm.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/tests/test_cephadm.py:2278]: just that the error messages end up formatted properly. It's the matching of the string against the error messages that's of value (since all the `async_timeout_handler` does is format error messages anyway)

File: src/pybind/mgr/cephadm/serve.py [hunk 2]
```diff
@@ -1459,6 +1460,11 @@ async def _run_cephadm(self,
         if not self.mgr.cgroups_split:
             final_args += ['--no-cgroups-split']
 
+        if not timeout:
+            # 15 minute global timeout if no timeout was passed
+            timeout = self.mgr.default_cephadm_command_timeout
+        final_args += ['--timeout', str(timeout)]
+
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/serve.py:1478]: I'm a bit iffy on the idea that the timeout passed to cephadm (binary) is the same timeout used in the mgr module. I'd prefer there to be a small "grace period" to allow the opportunity for the lower layer to exit and return and have it's result(s) be processed rather than "simultaneously" terminating the command and giving up on it.  This grace period could be only a few seconds. To have fewer touch points in the code this could just subtract, say 5, from the timeout before passing that adjusted timeout to cephadm.

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -686,8 +697,34 @@ def serve(self) -> None:
         serve = CephadmServe(self)
         serve.serve()
 
-    def wait_async(self, coro: Awaitable[T]) -> T:
-        return self.event_loop.get_result(coro)
+    def wait_async(self, coro: Awaitable[T], timeout: Optional[int] = None) -> T:
+        if not timeout:
+            timeout = self.default_cephadm_command_timeout
+        return self.event_loop.get_result(coro, timeout)
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/module.py:703]: FWIW, I prefer the current version as it leads to shorter lines that have fewer "things going on".

File: src/pybind/mgr/cephadm/upgrade.py [hunk 4]
```diff
@@ -174,7 +174,7 @@ def upgrade_status(self) -> orchestrator.UpgradeStatusSpec:
             # accessing self.upgrade_info_str will throw an exception if it
             # has not been set in _do_upgrade yet
             try:
-                r.message = self.upgrade_info_str
+                r.message = self.upgrade_info_str  # type: ignore
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/upgrade.py:177]: Why did you need to add `type: ignore` to this line?

**Review by phlogistonjohn — COMMENTED** [review]
No blockers, a few comments, but since I have a design q I'll leave this as a comment only: see my concern about having the mgr module timeout equal to the cephadm command timeout (as we've discussed previously). Feel free to tell me that is a concern that can be revisited later if you want.

File: src/pybind/mgr/cephadm/upgrade.py (hunk 4 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/upgrade.py]: For whatever reason, mypy suddenly noticed (and this was before your mypy version bump) that the variable didn't have a type
```
cephadm/upgrade.py: note: In member "upgrade_status" of class "CephadmUpgrade":
cephadm/upgrade.py:177: error: Cannot determine type of "upgrade_info_str"
```
I've just set the type in the class initialization now instead of the ignore, but still don't understand what changes I made here that made mypy become aware of it suddenly

File: src/pybind/mgr/cephadm/serve.py (hunk 2 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/serve.py]: I think that makes sense and this was an oversight while I was focused on just the asyncio timeout. I added the -5 here and also put a 60 second lower bound on the default while I was looking at it (although that's arguably too low, but better than nothing)

**Review by phlogistonjohn — APPROVED** [review]
Thanks for responding to my concerns, I had one little follow on, but otherwise it LGTM

File: src/pybind/mgr/cephadm/serve.py [hunk 5]
```diff
@@ -1461,8 +1472,18 @@ async def _run_cephadm(self,
...
 
         if not timeout:
-            # 15 minute global timeout if no timeout was passed
+            # default global timeout if no timeout was passed
             timeout = self.mgr.default_cephadm_command_timeout
+            # put a lower bound of 60 seconds in case users
+            # accidentally set it to something unreasonable.
+            # For example if they though it was in minutes
+            # rather than seconds
+            if timeout < 60:
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/serve.py:1481]: added an info log for when we set it to 60 because it's too low

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-04-05_05:00:01-orch:cephadm-wip-adk-testing-2023-04-04-2123-distro-default-smithi/

Reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2023-04-07_13:59:11-orch:cephadm-wip-adk-testing-2023-04-04-2123-distro-default-smithi/

After reruns: 6 failed and 2 dead jobs

- 2 dead jobs are an issue with the workload test after the upgrade has completed timing out. This happened in other runs on both quincy and reef done at the same time (with totally separate sets of patches), so I think it's something outside of ceph and not a blocker for merging PRs
- 1 failure is the new keepalive-nfs test introduced in a PR in the run that doesn't work yet
- 2 failures on staggered upgrade. Originally broken by https://tracker.ceph.com/issues/58535 and a PR in the run attempted to fix it, but the fix is not working yet
- 2 infra failures, one pulling the container image needed for the test and another installing the kernel
- 1 failure on testing jaegar tracing stuff.  It seems the jaegar-agent daemons weren't coming up. Will need to investigate this further, but I don't think it's related to any PRs in the run and shouldn't block merging.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-04-20_12:02:37-orch:cephadm-wip-adk-testing-2023-04-20-0105-distro-default-smithi/

Reruns of all but upgrade-with-workload jobs (which are known to timeout after multiple hours currently, so reruns would waste a lot of resources): https://pulpito.ceph.com/adking-2023-04-22_17:05:01-orch:cephadm-wip-adk-testing-2023-04-20-0105-distro-default-smithi/

After reruns, 2 dead and 5 failed jobs:

- 2 dead jobs were infra failures: `"Failed to download metadata for repo 'CentOS-PowerTools'`
- 2 failed jobs on staggered upgrade tests, https://tracker.ceph.com/issues/58535 + fix for it not working yet
- 2 failed jobs on new keepalive-only test. It's a new test that doesn't work yet, shouldn't affect being able to merge other PRs
- 1 failed test deploying jeagar-tracing. Needs investigation, but doesn't seem related to any PRs in the run.
